File: 2025-06-June-GFG-POTD/June 21 - Police and Thieves/Solution.java
class Solution {
   public int catchThieves(char[] arr, int k) {
        int n = arr.length;
        int count = 0;

        Queue<Integer> police = new LinkedList<>();
        Queue<Integer> thief = new LinkedList<>();

        for (int i = 0; i < n; i++) {
            if (arr[i] == 'P') {
                police.add(i);
            } else if (arr[i] == 'T') {
                thief.add(i);
            }

            while (!police.isEmpty() && !thief.isEmpty()) {
                int polIndex = police.peek();
                int thiefIndex = thief.peek();

                if (Math.abs(polIndex - thiefIndex) <= k) {
                    count++;
                    police.poll();
                    thief.poll();
                } else if (thiefIndex < polIndex) {
                    thief.poll();
                } else {
                    police.poll();
                }
            }
        }

        return count;
    }
}

File: 2025-06-June-GFG-POTD/June 21 - Police and Thieves/Solution.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

class Solution {
public:
    int catchThieves(vector<char>& arr, int k) {
        int n = arr.size();
        int count = 0;

        queue<int> police;
        queue<int> thief;

        for (int i = 0; i < n; ++i) {
            if (arr[i] == 'P') {
                police.push(i);
            } else if (arr[i] == 'T') {
                thief.push(i);
            }

            while (!police.empty() && !thief.empty()) {
                int polIndex = police.front();
                int thiefIndex = thief.front();

                if (abs(polIndex - thiefIndex) <= k) {
                    count++;
                    police.pop();
                    thief.pop();
                } else if (thiefIndex < polIndex) {
                    thief.pop();
                } else {
                    police.pop();
                }
            }
        }

        return count;
    }
};
